crml/nft/src/lib.rs

```diff
@@ -0,0 +1,651 @@
+			} else {
+				Zero::zero()
+			}
+		}
+
+		/// Create a new NFT collection
+		/// The caller will be come the collection' owner
+		/// `collection_id`- 32 byte utf-8 string
+		/// `schema` - for the collection
+		/// `royalties_schedule` - defacto royalties plan for secondary sales, this will apply to all tokens in the collection by default.
+		#[weight = T::WeightInfo::create_collection()]
+		fn create_collection(origin, collection_id: CollectionId, schema: NFTSchema, royalties_schedule: Option<RoyaltiesSchedule<T::AccountId>>) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+
+			ensure!(!collection_id.is_empty() && collection_id.len() <= MAX_COLLECTION_ID_LENGTH as usize, Error::<T>::CollectionIdInvalid);
+			ensure!(core::str::from_utf8(&collection_id).is_ok(), Error::<T>::CollectionIdInvalid);
+			ensure!(!<CollectionOwner<T>>::contains_key(&collection_id), Error::<T>::CollectionIdExists);
+
+			ensure!(!schema.is_empty(), Error::<T>::SchemaEmpty);
+			ensure!(schema.len() <= MAX_SCHEMA_FIELDS as usize, Error::<T>::SchemaMaxAttributes);
+
+			// Check the provided attribute types are valid
+			ensure!(
+				schema.iter().all(|(_name, type_id)| NFTAttributeValue::is_valid_type_id(*type_id)),
+				Error::<T>::SchemaInvalid
+			);
+
+			// Attribute names must be unique (future proofing for map lookups etc.)
+			let (attribute_names, _): (Vec<NFTAttributeName>, Vec<NFTAttributeTypeId>) = schema.iter().cloned().unzip();
+			let deduped = BTreeSet::from_iter(attribute_names);
+			ensure!(deduped.len() == schema.len(), Error::<T>::SchenmaDuplicateAttribute);
+
+			// Create the collection, update ownership, and bookkeeping
+			if let Some(royalties_schedule) = royalties_schedule {
+				ensure!(royalties_schedule.validate(), Error::<T>::RoyaltiesOvercommitment);
+				<CollectionRoyalties<T>>::insert(&collection_id, royalties_schedule);
+			}
+			CollectionSchema::insert(&collection_id, schema);
+			<CollectionOwner<T>>::insert(&collection_id, origin.clone());
+
+			Self::deposit_event(RawEvent::CreateCollection(collection_id, origin));
+
+			Ok(())
+		}
+
+		/// Issue a new NFT
+		/// `owner` - the token owner
+		/// `attributes` - initial values according to the NFT collection/schema
+		/// `royalties_schedule` - optional royalty schedule for secondary sales of _this_ token, defaults to the collection config
+		/// Caller must be the collection owner
+		#[weight = T::WeightInfo::create_token()]
+		#[transactional]
+		fn create_token(origin, collection_id: CollectionId, owner: T::AccountId, attributes: Vec<NFTAttributeValue>, royalties_schedule: Option<RoyaltiesSchedule<T::AccountId>>) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+
+			// Permission and existence check
+			let collection_owner = Self::collection_owner(&collection_id);
+			ensure!(collection_owner.is_some(), Error::<T>::NoCollection);
+			ensure!(collection_owner.unwrap() == origin, Error::<T>::NoPermission);
```

Comment:
If you do `ensure!(collection_owner == Some(origin), Error::<T>::NoPermission);` here, the safety of this line wouldn't depend on the previous line.

Reply:
it's like this to give a more specific error to the user depending on the situation

Reply:
That's alright to have second more specific error. What I am suggesting is that try to avoid `unwrap` which is not safe in general. Here it is safe because in the previous line you have checked `is_some`. But generally better not to be dependent on a previous line if there is a simpler way like what I said `ensure!(collection_owner == Some(origin), Error::<T>::NoPermission);`